test/object-store/c_api/c_api.cpp

```diff
@@ -2328,6 +2328,126 @@ TEST_CASE("C API") {
         }
     }
 
+    SECTION("freeze and thaw") {
+        SECTION("realm") {
+            auto frozen_realm = cptr_checked(realm_freeze(realm));
+            CHECK(!realm_is_frozen(realm));
+            CHECK(realm_is_frozen(frozen_realm.get()));
+
+            auto thawed_realm = cptr_checked(realm_thaw(frozen_realm.get()));
+            CHECK(realm_is_frozen(frozen_realm.get()));
+            CHECK(!realm_is_frozen(thawed_realm.get()));
+        }
+
+        SECTION("objects") {
+            CPtr<realm_object_t> obj1;
+            realm_value_t value;
+
+            write([&]() {
+                obj1 = cptr_checked(realm_object_create(realm, class_foo.key));
+                CHECK(obj1);
+            });
+            CHECK(checked(realm_get_value(obj1.get(), foo_str_key, &value)));
+            CHECK(value.type == RLM_TYPE_STRING);
+            CHECK(strncmp(value.string.data, "", value.string.size) == 0);
+
+            auto frozen_realm = cptr_checked(realm_freeze(realm));
+            auto frozen_obj1 = cptr_checked(realm_object_freeze(obj1.get(), frozen_realm.get()));
+            CHECK(frozen_obj1);
+
+            write([&]() {
+                CHECK(checked(realm_set_value(obj1.get(), foo_str_key, rlm_str_val("Hello, World!"), false)));
+            });
+
+            CHECK(checked(realm_get_value(obj1.get(), foo_str_key, &value)));
+            CHECK(value.type == RLM_TYPE_STRING);
+            CHECK(strncmp(value.string.data, "Hello, World!", value.string.size) == 0);
+
+            CHECK(checked(realm_get_value(frozen_obj1.get(), foo_str_key, &value)));
+            CHECK(value.type == RLM_TYPE_STRING);
+            CHECK(strncmp(value.string.data, "", value.string.size) == 0);
+
+            auto thawed_obj1 = cptr_checked(realm_object_thaw(frozen_obj1.get(), realm));
+            CHECK(thawed_obj1);
+            CHECK(checked(realm_get_value(thawed_obj1.get(), foo_str_key, &value)));
+            CHECK(value.type == RLM_TYPE_STRING);
+            CHECK(strncmp(value.string.data, "Hello, World!", value.string.size) == 0);
+        }
+
+        SECTION("results") {
```

Comment:
The only concern I have about this method is that there are right now 7 different ways to create a `Results` in ObjectStore and this test just test one (albeit the primary one by far), but I don't know if it is worth testing the rest (nor if it is possible from the C API).